File: apps/web/lib/order-protection-api.ts
import { getAuthed, postAuthed } from './authed-api';

const uuidPattern = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export interface ProtectionCaseRow {
  id: string;
  dispute_id: string;
  order_id: string;
  claimant_user_id: string;
  respondent_user_id: string;
  beneficiary_role: 'buyer' | 'seller';
  claim_type: string;
  policy_version: string;
  eligibility_basis: unknown;
  created_at: string;
  updated_at: string;
}

export interface OrderReturnRow {
  id: string;
  protection_case_id: string;
  dispute_id: string;
  order_id: string;
  buyer_id: string;
  seller_id: string;
  status: 'authorized' | 'awaiting_shipment' | 'in_transit' | 'delivered' | 'received' | 'contested' | 'resolved' | 'expired' | 'cancelled';
  reason: string;
  carrier: string | null;
  tracking_reference: string | null;
  tracking_url: string | null;
  return_due_at: string;
  shipped_at: string | null;
  delivered_at: string | null;
  received_at: string | null;
  seller_condition: 'accepted' | 'contested' | null;
  seller_condition_note: string | null;
  resolved_at: string | null;
}

export interface OrderProtectionSnapshot {
  cases: ProtectionCaseRow[];
  returns: OrderReturnRow[];
}

function id(value: string, label: string): string {
  const normalized = value.trim();
  if (!uuidPattern.test(normalized)) throw new Error(`${label} must be a UUID`);
  return normalized;
}

export function readOrderProtection(orderId: string): Promise<OrderProtectionSnapshot> {
  return getAuthed(`/v1/market/orders/${id(orderId, 'Order identifier')}/protection`);
}

export function shipProtectedReturn(returnId: string, input: {
  carrier: string;
  trackingReference: string;
  trackingUrl?: string;
}) {
  const carrier = input.carrier.trim();
  const trackingReference = input.trackingReference.trim();
  if (carrier.length < 2 || carrier.length > 80) throw new Error('Carrier must be 2-80 characters');
  if (trackingReference.length < 2 || trackingReference.length > 200) throw new Error('Tracking reference must be 2-200 characters');
  const trackingUrl = input.trackingUrl?.trim();
  if (trackingUrl) {
    const parsed = new URL(trackingUrl);
    if (parsed.protocol !== 'https:') throw new Error('Tracking URL must use HTTPS');
  }
  return postAuthed(`/v1/market/returns/${id(returnId, 'Return identifier')}/ship`, {
    carrier,
    trackingReference,
    ...(trackingUrl ? { trackingUrl } : {})
  });
}

export function acknowledgeProtectedReturn(returnId: string, input: {
  condition: 'accepted' | 'contested';
  note?: string;
}) {
  const note = input.note?.trim();
  if (input.condition === 'contested' && (!note || note.length < 8)) {
    throw new Error('A contest note of at least 8 characters is required');
  }
  return postAuthed(`/v1/market/returns/${id(returnId, 'Return identifier')}/receipt`, {
    condition: input.condition,
    ...(note ? { note } : {})
  });
}

File: apps/mobile/lib/src/api/order_protection_api.dart
import 'authed_api.dart';

final _uuidPattern = RegExp(r'^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[1-5][0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$');
const _returnStatuses = {
  'authorized',
  'awaiting_shipment',
  'in_transit',
  'delivered',
  'received',
  'contested',
  'resolved',
  'expired',
  'cancelled',
};

class MobileProtectionCase {
  const MobileProtectionCase({required this.id, required this.claimType, required this.policyVersion});
  final String id;
  final String claimType;
  final String policyVersion;

  factory MobileProtectionCase.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final claimType = json['claim_type'];
    final policyVersion = json['policy_version'];
    if (id is! String || !_uuidPattern.hasMatch(id) || claimType is! String || claimType.isEmpty || policyVersion is! String || policyVersion.isEmpty) {
      throw const FormatException('Invalid protection case');
    }
    return MobileProtectionCase(id: id, claimType: claimType, policyVersion: policyVersion);
  }
}

class MobileOrderReturn {
  const MobileOrderReturn({
    required this.id,
    required this.buyerId,
    required this.sellerId,
    required this.status,
    required this.reason,
    required this.returnDueAt,
    required this.carrier,
    required this.trackingReference,
    required this.trackingUrl,
    required this.sellerCondition,
    required this.sellerConditionNote,
  });

  final String id;
  final String buyerId;
  final String sellerId;
  final String status;
  final String reason;
  final DateTime returnDueAt;
  final String? carrier;
  final String? trackingReference;
  final String? trackingUrl;
  final String? sellerCondition;
  final String? sellerConditionNote;

  factory MobileOrderReturn.fromJson(Map<String, dynamic> json) {
    final id = json['id'];
    final buyerId = json['buyer_id'];
    final sellerId = json['seller_id'];
    final status = json['status'];
    final reason = json['reason'];
    final due = DateTime.tryParse(json['return_due_at']?.toString() ?? '');
    if (id is! String || !_uuidPattern.hasMatch(id) || buyerId is! String || !_uuidPattern.hasMatch(buyerId) || sellerId is! String || !_uuidPattern.hasMatch(sellerId) || status is! String || !_returnStatuses.contains(status) || reason is! String || due == null) {
      throw const FormatException('Invalid order return');
    }
    final trackingUrl = json['tracking_url'];
    if (trackingUrl != null) {
      final parsed = Uri.tryParse(trackingUrl.toString());
      if (parsed == null || parsed.scheme != 'https' || parsed.host.isEmpty) {
        throw const FormatException('Invalid return tracking URL');
      }
    }
    return MobileOrderReturn(
      id: id,
      buyerId: buyerId,
      sellerId: sellerId,
      status: status,
      reason: reason,
      returnDueAt: due,
      carrier: json['carrier']?.toString(),
      trackingReference: json['tracking_reference']?.toString(),
      trackingUrl: trackingUrl?.toString(),
      sellerCondition: json['seller_condition']?.toString(),
      sellerConditionNote: json['seller_condition_note']?.toString(),
    );
  }
}

class MobileOrderProtection {
  const MobileOrderProtection({required this.cases, required this.returns});
  final List<MobileProtectionCase> cases;
  final List<MobileOrderReturn> returns;

  factory MobileOrderProtection.fromJson(Map<String, dynamic> json) {
    final casesRaw = json['cases'];
    final returnsRaw = json['returns'];
    if (casesRaw is! List || returnsRaw is! List) throw const FormatException('Invalid order protection snapshot');
    return MobileOrderProtection(
      cases: casesRaw.map((row) {
        if (row is! Map) throw const FormatException('Invalid protection case row');
        return MobileProtectionCase.fromJson(Map<String, dynamic>.from(row));
      }).toList(growable: false),
      returns: returnsRaw.map((row) {
        if (row is! Map) throw const FormatException('Invalid order return row');
        return MobileOrderReturn.fromJson(Map<String, dynamic>.from(row));
      }).toList(growable: false),
    );
  }
}

abstract interface class OrderProtectionGateway {
  Future<MobileOrderProtection> read(String accessToken, String orderId);
  Future<void> shipReturn(String accessToken, {required String returnId, required String carrier, required String trackingReference, String? trackingUrl});
  Future<void> acknowledgeReturn(String accessToken, {required String returnId, required String condition, String? note});
}

class OrderProtectionApi implements OrderProtectionGateway {
  const OrderProtectionApi({required AuthedApi authedApi}) : _api = authedApi;
  final AuthedApi _api;

  String _id(String value) {
    final result = value.trim();
    if (!_uuidPattern.hasMatch(result)) throw const FormatException('Invalid identifier');
    return result;
  }

  @override
  Future<MobileOrderProtection> read(String accessToken, String orderId) async {
    return MobileOrderProtection.fromJson(await _api.get('/v1/market/orders/${_id(orderId)}/protection', accessToken));
  }

  @override
  Future<void> shipReturn(String accessToken, {required String returnId, required String carrier, required String trackingReference, String? trackingUrl}) async {
    final carrierValue = carrier.trim();
    final trackingValue = trackingReference.trim();
    if (carrierValue.length < 2 || carrierValue.length > 80 || trackingValue.length < 2 || trackingValue.length > 200) {
      throw const FormatException('Invalid return shipment');
    }
    final urlValue = trackingUrl?.trim();
    if (urlValue != null && urlValue.isNotEmpty) {
      final parsed = Uri.tryParse(urlValue);
      if (parsed == null || parsed.scheme != 'https' || parsed.host.isEmpty) throw const FormatException('Tracking URL must use HTTPS');
    }
    await _api.post('/v1/market/returns/${_id(returnId)}/ship', accessToken, {
      'carrier': carrierValue,
      'trackingReference': trackingValue,
      if (urlValue != null && urlValue.isNotEmpty) 'trackingUrl': urlValue,
    });
  }

  @override
  Future<void> acknowledgeReturn(String accessToken, {required String returnId, required String condition, String? note}) async {
    if (condition != 'accepted' && condition != 'contested') throw const FormatException('Invalid return condition');
    final noteValue = note?.trim();
    if (condition == 'contested' && (noteValue == null || noteValue.length < 8 || noteValue.length > 4000)) {
      throw const FormatException('Contest note must be 8-4000 characters');
    }
    await _api.post('/v1/market/returns/${_id(returnId)}/receipt', accessToken, {
      'condition': condition,
      if (noteValue != null && noteValue.isNotEmpty) 'note': noteValue,
    });
  }
}
